source/common/router/config_impl.cc

```diff
@@ -156,6 +144,52 @@ Upstream::RetryPrioritySharedPtr RetryPolicyImpl::retryPriority() const {
                                                            *validation_visitor_, num_retries_);
 }
 
+InternalRedirectPolicyImpl::InternalRedirectPolicyImpl(
+    const envoy::config::route::v3::InternalRedirectPolicy& policy_config,
+    ProtobufMessage::ValidationVisitor& validator, absl::string_view current_route_name)
+    : current_route_name_(current_route_name),
+      redirect_response_codes_(buildRedirectResponseCodes(policy_config)),
+      max_internal_redirects_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(policy_config, max_internal_redirects, 1)),
+      enabled_(true),
+      disallow_cross_scheme_redirect_(policy_config.disallow_cross_scheme_redirect()) {
+  for (const auto& predicate : policy_config.predicates()) {
+    const std::string type{TypeUtil::typeUrlToDescriptorFullName(predicate.type_url())};
+    auto* factory =
+        Registry::FactoryRegistry<InternalRedirectPredicateFactory>::getFactoryByType(type);
+
+    auto config = factory->createEmptyConfigProto();
+    Envoy::Config::Utility::translateOpaqueConfig(predicate, {}, validator, *config);
+    predicate_factories_.emplace_back(factory, std::move(config));
+  }
+}
+
+std::vector<InternalRedirectPredicateSharedPtr> InternalRedirectPolicyImpl::predicates() const {
+  std::vector<InternalRedirectPredicateSharedPtr> predicates;
+  for (const auto& predicate_factory : predicate_factories_) {
+    predicates.emplace_back(predicate_factory.first->createInternalRedirectPredicate(
+        *predicate_factory.second, current_route_name_));
+  }
+  return predicates;
+}
+
+absl::flat_hash_set<Http::Code> InternalRedirectPolicyImpl::buildRedirectResponseCodes(
+    const envoy::config::route::v3::InternalRedirectPolicy& policy_config) const {
+  if (policy_config.redirect_response_codes_size() == 0) {
+    return absl::flat_hash_set<Http::Code>{Http::Code::Found};
+  }
+  absl::flat_hash_set<Http::Code> ret;
+  std::for_each(policy_config.redirect_response_codes().begin(),
+                policy_config.redirect_response_codes().end(), [&ret](uint32_t response_code) {
+                  const absl::flat_hash_set<uint32_t> valid_redirect_response_code = {301, 302, 303,
```

Comment:
Do we test invalid config to make sure it's dropped?

Reply:
Done.